internal/services/rate_test.go

```diff
@@ -13,46 +14,70 @@ const (
 	TestingRateServiceWaitingDuration = time.Second * 2
 )
 
-type MockExchangeRateClient struct {
+type MockExchangeRateFetcher struct {
 	mock.Mock
 }
 
-func (ec *MockExchangeRateClient) GetExchangeRate() (float32, error) {
+func (ec *MockExchangeRateFetcher) FirstFetcher(code string, client *http.Client) (float32, error) {
+	args := ec.Called()
+	return args.Get(0).(float32), args.Error(1)
+}
+
+func (ec *MockExchangeRateFetcher) SecondFetcher(code string, client *http.Client) (float32, error) {
 	args := ec.Called()
 	return args.Get(0).(float32), args.Error(1)
 }
 
 func TestRateService_RateIsCorrect(t *testing.T) {
-	mockRateClient := new(MockExchangeRateClient)
-	mockRateClient.On("GetExchangeRate").Return(float32(8.0), nil)
-	rateService := NewRateService(TestingRateServiceFetchInterval, mockRateClient)
-	rateService.StartBackgroundTask()
+	mockRateFetcher := new(MockExchangeRateFetcher)
+	mockRateFetcher.On("FirstFetcher").Return(float32(8.0), nil)
+
+	rateService := NewRateService(WithFetchers(CreateNamedFetcher(mockRateFetcher.FirstFetcher, "first")))
+	rate, err := rateService.GetRate("usd")
 
-	rate, err := rateService.GetRate()
 	assert.Equal(t, float32(8.0), rate)
 	assert.NoError(t, err)
 }
 
 func TestRateService_RateIsCached(t *testing.T) {
-	mockRateClient := new(MockExchangeRateClient)
-	mockRateClient.On("GetExchangeRate").Return(float32(8.0), nil)
-	rateService := NewRateService(TestingRateServiceFetchInterval, mockRateClient)
-	rateService.StartBackgroundTask()
-
-	_, _ = rateService.GetRate()
-	_, _ = rateService.GetRate()
+	mockRateFetcher := new(MockExchangeRateFetcher)
+	mockRateFetcher.On("FirstFetcher").Return(float32(8.0), nil)
+	rateService := NewRateService(WithFetchers(CreateNamedFetcher(mockRateFetcher.FirstFetcher, "first")))
+	_, _ = rateService.GetRate("usd")
+	_, _ = rateService.GetRate("usd")
 
 	// Make sure service cached result from previoues calls.
-	mockRateClient.AssertNumberOfCalls(t, "GetExchangeRate", 1)
+	mockRateFetcher.AssertNumberOfCalls(t, "FirstFetcher", 1)
 }
 
 func TestRateService_RateIsFetchedAfterInterval(t *testing.T) {
-	mockRateClient := new(MockExchangeRateClient)
-	mockRateClient.On("GetExchangeRate").Return(float32(8.0), nil)
-	rateService := NewRateService(TestingRateServiceFetchInterval, mockRateClient)
-	rateService.StartBackgroundTask()
+	mockRateFetcher := new(MockExchangeRateFetcher)
+	mockRateFetcher.On("FirstFetcher").Return(float32(8.0), nil)
+	rateService := NewRateService(
+		WithFetchers(CreateNamedFetcher(mockRateFetcher.FirstFetcher, "first")),
+		WithUpdateInterval(TestingRateServiceFetchInterval))
 
 	// Make sure service re-fetch after update interval.
+	_, _ = rateService.GetRate("usd")
 	time.Sleep(TestingRateServiceWaitingDuration)
-	assert.GreaterOrEqual(t, len(mockRateClient.Calls), 2)
+	_, _ = rateService.GetRate("usd")
+
+	mockRateFetcher.AssertNumberOfCalls(t, "FirstFetcher", 2)
+}
+
+func TestRateService_FallbackToAnotherFetcher(t *testing.T) {
```

Comment:
Unfortunately, this test doesn't pass successfully(
![image](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/assets/24816241/6e777e8d-4954-407c-bca5-dc3628cfcc72)


I suppose the root cause is the way [`WithFetchers`](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/5/files#diff-5ed4d654fee99e78ea02f2a6c861077af36b53073891be54b973af158d060c8bR122-R125) option is implemented. Could you please fix it?

Reply:
Yeah, sure. Fixed!